### PR TITLE
feat: include replicas in volume snapshot recovery

### LINKED ISSUE #2947: Include replicas in volume snapshot recovery
**Premise:** we knew we would get to this point / great we are able to fix this just by leveraging Kubernetes storage resources.

While testing volume snapshot recovery with medium (500GB) and large size databases (4TB), as expected the recovery time to spin up the primary reduces dramatically (I am working on a blog post to cover this). However, as expected, replicas are then created using the default available method at the moment: pg_basebackup.

We should change the replica creation method to use the same volume snapshots set to spin them up.

As you can imagine, this possibility opens up a tons of interesting scenarios that gravitate around using volume snapshots and PVC cloning in general as a way to generate replicas, not only when boostrapping from a volume snapshot backup (i.e. a replica dies we could use the latest snapshot provided WAL files exist, or the whole cluster dies, we could spin it up from the last available snapshot and use the WAL archive to self-heal). All these scenarios are premature now.

However, if possible, we should investigate and add the possibility to at least build the replicas in the case of bootstrap recovery from a volume snapshot. If we see that this patch can land, we should add it to 1.21.

### PR DESCRIPTION
**Pull request by leonardoce**
When bootstrapping a cluster from a volume snapshot backup,
reuse the volume snapshot to build the replicas.

Closes #2947

### CODE REVIEW

**Comment by gbartolini** [comment]
I ❤️ this!

**Review by armru — APPROVED** [review]

**Comment by leonardoce** [comment]
Thank you @armru for your review. I'd like to change a few more things in this PR so I'm assigning it to me.

**Comment by leonardoce** [comment]
@NiccoloFei, unfortunately, I had to remove the improvement you did in the E2e test since now we (correctly) require WAL archiving to be active when we bootstrap replicas using snapshots.
Perhaps we could create an E2e test just for this scenario?

File: pkg/reconciler/persistentvolumeclaim/storagesource_test.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+					Phase: apiv1.BackupPhaseCompleted,
+				},
+			},
+		},
+	}
+
+	When("bootstrapping from a VolumeSnapshot", func() {
+		When("we don't have backups", func() {
+			When("there's no source WAL archive", func() {
+				It("should fail when looking for a wrong role", func(ctx context.Context) {
```
Comment (jsilvela) [inline pkg/reconciler/persistentvolumeclaim/storagesource_test.go:122]: oooh, nice! Very readable. I had never thought of nesting When's

File: tests/utils/job.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"errors"
+	"fmt"
+
+	batchv1 "k8s.io/api/batch/v1"
+)
+
+// GetJob gets a Job by namespace and name
+func (env TestingEnvironment) GetJob(namespace, jobName string) (*batchv1.Job, error) {
+	wrapErr := func(err error) error {
```
Comment (jsilvela) [inline tests/utils/job.go:28]: This function seems unused. Still, good idea to have it, and the linter is not complaining...

**Review by jsilvela — APPROVED** [review]

File: tests/utils/job.go (hunk 2 above)
Comment (leonardoce) [reply tests/utils/job.go:28]: I think this will be needed when we add E2e tests for replica cloning via Volume Snapshots.

**Comment by gbartolini** [comment]
We should IMHO have documented this behaviour.

**Pull request merged**